app/models/publishing_pipeline_state.rb

```diff
@@ -0,0 +1,181 @@
+    publishing_queue_item.update!(last_pipeline_state: status)
+  end
+
+  def no_update
+    errors.add(:base, "cannot update!")
+  end
+
+  def podcast_ids_match
+    if podcast_id != publishing_queue_item&.podcast_id
+      errors.add(:podcast_id, "must match the podcast_id of the publishing_queue_item")
+    end
+  end
+
+  def no_transition_from_terminal_state
+    if done?
+      errors.add(:status, "cannot transition from a terminal state")
+    end
+  end
+
+  def self.terminal_status_codes
+    TERMINAL_STATUSES.map { |s| statuses[s] }
+  end
+
+  def self.unique_status_codes
+    UNIQUE_STATUSES.map { |s| statuses[s] }
+  end
+
+  def self.most_recent_state(podcast)
+    latest_by_podcast.where(podcast_id: podcast.id).first
+  end
+
+  def self.start_pipeline!(podcast)
+    PublishingQueueItem.ensure_queued!(podcast)
+    attempt!(podcast)
+  end
+
+  # None of the methods that grab locks are threadsafe if we assume that
+  # creating published artifacts is non-idempotent (e.g. creating remote Apple
+  # resources)
+  def self.attempt!(podcast, perform_later: true)
+    podcast.with_publish_lock do
+      next if PublishingQueueItem.unfinished_items(podcast).empty?
+      next if PublishingQueueItem.current_unfinished_item(podcast).present?
+
+      # Dedupe the work, grab the latest unfinished item in the queue
+      latest_unfinished_item = PublishingQueueItem.unfinished_items(podcast).first
+
+      PublishingPipelineState.create!(podcast: podcast, publishing_queue_item: latest_unfinished_item, status: :created)
+
+      if perform_later
+        Rails.logger.info("Scheduling PublishFeedJob for podcast #{podcast.id}", {podcast_id: podcast.id})
+        PublishFeedJob.perform_later(podcast)
+      else
+        Rails.logger.info("Performing PublishFeedJob for podcast #{podcast.id}", {podcast_id: podcast.id})
+        PublishFeedJob.perform_now(podcast)
+      end
+    end
+  end
+
+  def self.expired?(podcast)
+    expired_pipelines.where(podcast: podcast).exists?
+  end
+
+  def self.start!(podcast)
+    state_transition(podcast, :started)
+  end
+
+  def self.publish_rss!(podcast)
+    state_transition(podcast, :published_rss)
+  end
+
+  def self.publish_apple!(podcast)
+    state_transition(podcast, :published_apple)
+  end
+
+  def self.complete!(podcast)
+    state_transition(podcast, :complete)
+  end
+
+  def self.error!(podcast)
+    state_transition(podcast, :error)
+  end
+
+  def self.expire!(podcast)
+    state_transition(podcast, :expired)
+  end
+
+  def self.expire_pipelines!
+    Podcast.where(id: expired_pipelines.select(:podcast_id)).each do |podcast|
+      Rails.logger.error("Cleaning up expired publishing pipeline for podcast #{podcast.id}", {podcast_id: podcast.id})
+      expire!(podcast)
+    end
+  end
+
+  def self.settle_remaining!(podcast)
+    attempt!(podcast)
+  end
+
+  def self.complete?(podcast)
+    most_recent_state(podcast)&.complete?
+  end
+
+  def self.state_transition(podcast, to_state)
```

Comment:
So the assumption is you'd never be doing a state transition for any PublishingPipelineState _other than_ the latest unfinished one?

Reply:
Right, and also that is per podcast.  Only one publishing pipeline can kicked off at a time.